### PR TITLE
Yif inspector view redesign

### LINKED ISSUE #470: Redesign element inspector sidebar
Make views for FunctionSegemtnBBMs

### LINKED ISSUE #476: Repeat count and Absolute Length params not saved upon enter
These params in the User Defined function only save if you use the arrows to click them, not when they are initially added by user.

### PR DESCRIPTION
**Pull request by i-m-foster**
Draft pull request. It's not done yet but I want to check on our progress and see if it has the correct intended behavior

### CODE REVIEW

**Comment by i-m-foster** [comment]
Also, the chart is not working yet

File: leaf-ui/css/style.css [hunk 1]
```diff
@@ -276,28 +276,74 @@ body.waiting *
...
 #assigned-vals{
-  background-color: #1E85F7;
+  background-color: rgba(36, 150, 255, 1);
   border-color: rgba(31, 121, 101, 1);
   color: white;
   font-size: 0.7em;
   font-weight: 200;
   width: 200px;
-  display: block;
+  display: block;
```
Comment (Domily11) [inline leaf-ui/css/style.css]: Take out extra space

File: leaf-ui/rappid-extensions/BackboneModelsIntention.js [hunk 2]
```diff
@@ -229,12 +232,10 @@ var IntentionBBM = Backbone.Model.extend({
     changeInitialSatValue: function(initValue) {
         // Set the the first element of userEvaluationList to the initValue
         this.getUserEvaluationBBM(0).set('assignedEvidencePair', initValue);;
-
```
Comment (Domily11) [inline leaf-ui/rappid-extensions/BackboneModelsIntention.js:232]: Add back space

File: leaf-ui/rappid-extensions/ElementInspector.js [hunk 3]
```diff
@@ -91,7 +83,7 @@ var ElementInspector = Backbone.View.extend({
                 '<option value="0100"> Partially Denied (⊥, P)</option>',
                 '<option value="1100"> Denied (⊥, F)</option>',
             '</select>',
-            '<br>',
+            // '<br>',
```
Comment (Domily11) [inline leaf-ui/rappid-extensions/ElementInspector.js]: Decide whether or not to keep

File: leaf-ui/rappid-extensions/ElementInspector.js [hunk 4]
```diff
@@ -163,10 +133,8 @@ var ElementInspector = Backbone.View.extend({
         'change #init-sat-value':'initSatValueChanged',
 
         'change .function-type':'funcTypeChanged',
-        'change .function-sat-value':'funcSatValChanged',
-
-        'change .user-function-type':'userFuncTypeChanged',
-        'change .user-sat-value':'userSatValChanged',
+        'change .segment-functions':'updateHTML',
+
```
Comment (Domily11) [inline leaf-ui/rappid-extensions/ElementInspector.js]: remove extra line

File: leaf-ui/rappid-extensions/ElementInspector.js [hunk 5]
```diff
@@ -183,7 +151,7 @@ var ElementInspector = Backbone.View.extend({
      */
     render: function() {
         this.$el.html(_.template($(this.template).html())(this.model.toJSON()))
-
```
Comment (Domily11) [inline leaf-ui/rappid-extensions/ElementInspector.js:186]: remove extra space

File: leaf-ui/rappid-extensions/ElementInspector.js [hunk 6]
```diff
@@ -202,6 +170,7 @@ var ElementInspector = Backbone.View.extend({
             this.model.attr(".funcvalue/text", ' ');
         }
         
+
```
Comment (Domily11) [inline leaf-ui/rappid-extensions/ElementInspector.js]: Remove extra line

File: leaf-ui/rappid-extensions/ElementInspector.js [hunk 7]
```diff
@@ -217,11 +186,8 @@ var ElementInspector = Backbone.View.extend({
                 this.updateHTML();
             }
             
-            if(functionType == 'I' || functionType == 'D' || functionType == 'MN' || functionType == 'MP'){
-                this.displayFunctionSatValue(null);
-            }
         }
-
+        // this.$('.inspector-views').append(this.innerView.$el);
```
Comment (Domily11) [inline leaf-ui/rappid-extensions/ElementInspector.js]: Take this out?

File: leaf-ui/rappid-extensions/ElementInspector.js [hunk 8]
```diff
@@ -270,103 +235,20 @@ var ElementInspector = Backbone.View.extend({
...
-                $(".user-sat-value").last().prop('disabled', true);
-                $(".user-sat-value").last().css("background-color",'grey');
-                $(".user-function-type").last().prop('disabled', true);
-                $(".user-function-type").last().css("background-color", 'grey');
-                html.appendTo(this.$('#all-user-constraints'));
-            }
-            })
+        // TODO: do we need this??
+        //     if (functionSegment !== funcSegments[len - 1]) {
+        //         // if it is not the last function segment, clone the select tags,
```
Comment (Domily11) [inline leaf-ui/rappid-extensions/ElementInspector.js]: Take this out?

File: leaf-ui/rappid-extensions/ElementInspector.js [hunk 9]
```diff
@@ -380,7 +262,7 @@ var ElementInspector = Backbone.View.extend({
             $("#repeat-end3").val(this.intention.get('evolvingFunction').get('repAbsTime'));
         }
 
-        this.updateChartUserDefined();  
+       // this.updateChartUserDefined();
```
Comment (Domily11) [inline leaf-ui/rappid-extensions/ElementInspector.js]: Take out

File: leaf-ui/rappid-extensions/ElementInspector.js [hunk 10]
```diff
@@ -452,35 +302,20 @@ var ElementInspector = Backbone.View.extend({
     updateHTML: function(event) {
         // Check if selected init sat value and functionType pair is illegal
         // Only runs if evolvingFunction is defined and therefore there is a function type
-        // if (this.intention.get('evolvingFunction') != null) {
```
Comment (Domily11) [inline leaf-ui/rappid-extensions/ElementInspector.js:455]: Do we not need this anymore?

File: leaf-ui/rappid-extensions/ElementInspector.js [hunk 11]
```diff
@@ -452,35 +302,20 @@ var ElementInspector = Backbone.View.extend({
...
-            this.validityCheck(event);
-        // }
+        this.validityCheck(event);
 
         if (this.intention.get('evolvingFunction') != null) {
             var functionType = this.intention.get('evolvingFunction').get('type');
+
```
Comment (Domily11) [inline leaf-ui/rappid-extensions/ElementInspector.js]: Remove extra line

File: leaf-ui/rappid-extensions/ElementInspector.js [hunk 12]
```diff
@@ -490,8 +325,8 @@ var ElementInspector = Backbone.View.extend({
                 this.$('#init-sat-value').prop('disabled', false);
             }
         }
-
-        this.updateChart(); 
+        this.rerender();
+      //  this.updateChart();
```
Comment (Domily11) [inline leaf-ui/rappid-extensions/ElementInspector.js]: Remove this

File: leaf-ui/rappid-extensions/ElementInspector.js [hunk 13]
```diff
@@ -529,300 +364,87 @@ var ElementInspector = Backbone.View.extend({
...
-        }
-        if (markedValue) {
-            if (satisfactionValuesDict[markedValue != null]){
-                this.$('#markedValue').val(satisfactionValuesDict[markedValue].name());
-            }
-        }
-        this.$('#markedValue').change();
-        return;
-    },
+
```
Comment (Domily11) [inline leaf-ui/rappid-extensions/ElementInspector.js:354]: Remove extra line

File: leaf-ui/rappid-extensions/ElementInspector.js [hunk 14]
```diff
@@ -529,300 +364,87 @@ var ElementInspector = Backbone.View.extend({
...
-        } else if (func == 'C') {
-            $(".user-sat-value").last().html(this.satValueOptionsAll());
-            // Restrict input if it is the first constraint
-            if (this.intention.getFuncSegments().length == 1) {
-                $(".user-sat-value").last().val(this.$('#init-sat-value').val())
-                $(".user-sat-value").last().prop('disabled', true);
-                $(".user-sat-value").last().css("background-color","grey");
-            }
-        } 
+        //this.updateChartUserDefined();
```
Comment (Domily11) [inline leaf-ui/rappid-extensions/ElementInspector.js]: Remove this

File: leaf-ui/rappid-extensions/ElementInspector.js [hunk 15]
```diff
@@ -529,300 +364,87 @@ var ElementInspector = Backbone.View.extend({
...
-        } else if (nextChar == end) {
-            $("#repeat-error").text("Repeated range must be at least two apart");
-            $("#repeat-error").show("fast");
-
-        } else {
-
-            $("#repeat-error").hide();
-            this.intention.get('evolvingFunction').setRepeatingFunction(begin, end, count, absTime);
-        }
-        this.updateChartUserDefined();
```
Comment (Domily11) [inline leaf-ui/rappid-extensions/ElementInspector.js:824]: Why move the order?

File: leaf-ui/rappid-extensions/ElementInspector.js [hunk 16]
```diff
@@ -529,300 +364,87 @@ var ElementInspector = Backbone.View.extend({
...
-
-            $("#repeat-error").hide();
-            this.intention.get('evolvingFunction').setRepeatingFunction(begin, end, count, absTime);
-        }
-        this.updateChartUserDefined(); 
-    },
+            $("#repeat-error").hide();
+            this.intention.get('evolvingFunction').setRepeatingFunction(begin, end, count, absTime);
+        }
+        //this.updateChartUserDefined();
```
Comment (Domily11) [inline leaf-ui/rappid-extensions/ElementInspector.js]: Delete

File: leaf-ui/rappid-extensions/ElementInspector.js [hunk 17]
```diff
@@ -838,7 +460,7 @@ var ElementInspector = Backbone.View.extend({
             $('#repeat-end2').val(2);
         }
         this.intention.get('evolvingFunction').set('repCount', repVal);
-        this.updateChartUserDefined(); 
+        //this.updateChartUserDefined();
```
Comment (Domily11) [inline leaf-ui/rappid-extensions/ElementInspector.js]: Delete

File: leaf-ui/rappid-extensions/ElementInspector.js [hunk 18]
```diff
@@ -853,7 +475,7 @@ var ElementInspector = Backbone.View.extend({
             $('#repeat-end3').val(0);
         }
         this.intention.get('evolvingFunction').set('repAbsTime', absLength);
-        this.updateChartUserDefined();
+        //this.updateChartUserDefined();
```
Comment (Domily11) [inline leaf-ui/rappid-extensions/ElementInspector.js]: Delete

File: leaf-ui/rappid-extensions/ElementInspector.js [hunk 19]
```diff
@@ -991,6 +613,473 @@ var ElementInspector = Backbone.View.extend({
...
+ 
+        // Sets the stopTP to be one step after the startTP
+        if (this.model.get('startTP') != '0') {
+            this.stopTP = String.fromCharCode(this.model.get('startTP').charCodeAt(0) + 1);
+        }
+        else {
+            this.stopTP = 'A';
+        }
+        // Boolean if the function segment is a part of a UD function
+        this.hasUD = options.hasUD;
```
Comment (Domily11) [inline leaf-ui/rappid-extensions/ElementInspector.js]: Don't need this

File: leaf-ui/rappid-extensions/ElementInspector.js [hunk 20]
```diff
@@ -991,6 +613,473 @@ var ElementInspector = Backbone.View.extend({
...
+        this.index = options.index;
+        this.initSatValue = options.initSatValue;
```
Comment (Domily11) [inline leaf-ui/rappid-extensions/ElementInspector.js:661]: Don't need to be passed in, can just directly get

File: leaf-ui/rappid-extensions/ElementInspector.js [hunk 21]
```diff
@@ -991,6 +613,473 @@ var ElementInspector = Backbone.View.extend({
...
+        }
+        
+        if (this.hasUD == true) {
+            this.checkUDFunctionValues()
+        }
+        else {
+            this.checkSatValue();
+        }
+    },
+
```
Comment (Domily11) [inline leaf-ui/rappid-extensions/ElementInspector.js:801]: Remove extra line

File: leaf-ui/rappid-extensions/ElementInspector.js [hunk 22]
```diff
@@ -991,6 +613,473 @@ var ElementInspector = Backbone.View.extend({
...
+     resetChartDatasets: function(datasets) {
+        for (var i = 0; i < datasets.length; i++) {
+            datasets[i].borderDash = [];
+            datasets[i].data = [];
+            datasets[i].pointBackgroundColor = ["rgba(220,220,220,1)", "rgba(220,220,220,1)", "rgba(220,220,220,1)"];
+            datasets[i].pointBorderColor = ["rgba(220,220,220,1)", "rgba(220,220,220,1)", "rgba(220,220,220,1)"];
+            datasets[i].borderColor = "rgba(220,220,220,1)";
+        }  
+    },
+
```
Comment (Domily11) [inline leaf-ui/rappid-extensions/ElementInspector.js:1083]: Remove extra line

File: leaf-ui/rappid-extensions/ElementInspector.js [hunk 23]
```diff
@@ -991,6 +613,473 @@ var ElementInspector = Backbone.View.extend({
...
+        var res = ['0'];
+        var curr = 'A'
+        for (var i = 0; i < num; i++) {
+            res.push(curr);
+            curr = String.fromCharCode(curr.charCodeAt(0) + 1);
+        }
+        return res;
+    },
+});
+
```
Comment (Domily11) [inline leaf-ui/rappid-extensions/ElementInspector.js]: remove extra line

File: leaf-ui/rappid-extensions/ElementInspector.js (hunk 10 above)
Comment (i-m-foster) [reply leaf-ui/rappid-extensions/ElementInspector.js:455]: Yeah, I think that was a quick fix for something so it's no longer needed

File: leaf-ui/rappid-extensions/BackboneModelsIntention.js [hunk 24]
```diff
@@ -277,6 +278,9 @@ var IntentionBBM = Backbone.Model.extend({
         var seg2 = null;
         // Creates the correct FunctionSegmentBBM(s) for the selected function type
         switch(funcType) {
+            case 'NT': 
+            var seg =  new FunctionSegmentBBM({type: funcType, refEvidencePair: '(no value)', startTP: '0', startAT: 0});
```
Comment (YesugenB) [inline leaf-ui/rappid-extensions/BackboneModelsIntention.js]: Fix the indent for var seg

File: leaf-ui/rappid-extensions/ElementInspector.js [hunk 25]
```diff
@@ -108,39 +100,17 @@ var ElementInspector = Backbone.View.extend({
...
-                            '<option value="1100"> Denied (⊥, F)</option>',
-                            '<option value="(no value)"> (no value) </option>',
-                            '</select>',
-                        '</div>',
-                    '</div>',
                     '<br>',
                     // Error message is controlled dynamically
                     '<label id="repeat-error"></label>',
-                    '<select id="repeat-begin" class="repeat-select">',
+                    '<select id="repeat-begin" class="repeat-select"style = "position:relative; left:38px; width: 93px">',
```
Comment (YesugenB) [inline leaf-ui/rappid-extensions/ElementInspector.js]: Space before style

File: leaf-ui/rappid-extensions/ElementInspector.js [hunk 26]
```diff
@@ -529,300 +364,87 @@ var ElementInspector = Backbone.View.extend({
...
-        var func = $(".user-function-type").last().val();
+    addSegment: function() {
+        // Adds a new FunctionSegmentBBM to the functionSegList
+        this.intention.addUserDefinedSeg("C", "0000");
+        var funcSegList = this.intention.getFuncSegments();
+        // Sets current to false so it becomes disabled for the previous FunctionSegmentBBM to 
+        if(funcSegList.length>1){
+            funcSegList[funcSegList.length-2].set('current', false);
+        }
+        // creates a new FunctionSegmentView from the new FunctionSegmentBBM
```
Comment (YesugenB) [inline leaf-ui/rappid-extensions/ElementInspector.js]: Capitalize `creates`

File: leaf-ui/rappid-extensions/ElementInspector.js [hunk 27]
```diff
@@ -529,300 +364,87 @@ var ElementInspector = Backbone.View.extend({
...
-            datasets[i].borderColor = "rgba(220,220,220,1)";
-        }  
+    repeatConstraintControl: function(){
+        if (!this.repeatOptionsDisplay){
+            this.setRepeatConstraintMode("TurnOn");
+            this.setRepeatConstraintMode("Update");
+        } else if (this.repeatOptionsDisplay){
+            this.setRepeatConstraintMode("TurnOff");
+            this.intention.get('evolvingFunction').removeRepFuncSegments();
+           // this.updateChartUserDefined();
```
Comment (YesugenB) [inline leaf-ui/rappid-extensions/ElementInspector.js]: Can this be yeeted

File: leaf-ui/rappid-extensions/ElementInspector.js [hunk 28]
```diff
@@ -991,6 +613,473 @@ var ElementInspector = Backbone.View.extend({
...
+                    var functionSegView = new FuncSegView({model: funcSeg, intention:this.intention, hasUD: hasUD, index: i, initSatValue: this.intention.getUserEvaluationBBM(0).get('assignedEvidencePair'), chart: this.chart});
+                    $('#segment-functions').append(functionSegView.el);
+                    functionSegView.render(); 
+                    i++; 
+            })
+        }
+    },
+
+    rerender: function() {
+        // Ads absTime label
```
Comment (YesugenB) [inline leaf-ui/rappid-extensions/ElementInspector.js]: Adds?

File: leaf-ui/rappid-extensions/ElementInspector.js [hunk 29]
```diff
@@ -991,6 +613,473 @@ var ElementInspector = Backbone.View.extend({
...
+    },
+});
+
+/************************************************** FunctionSegmentBBM View **************************************************/
+
+var FuncSegView = Backbone.View.extend({
+     model: FunctionSegmentBBM, 
+
+    initialize: function(options){ 
+        // Reference to the cahrt in the ElementInspector view
```
Comment (YesugenB) [inline leaf-ui/rappid-extensions/ElementInspector.js]: chart

File: leaf-ui/rappid-extensions/ElementInspector.js [hunk 30]
```diff
@@ -991,6 +613,473 @@ var ElementInspector = Backbone.View.extend({
...
+        this.model.set('startAT', absTime);         
+    },    
+
+    setFuncSatValue: function () {
+        this.model.set('refEvidencePair', this.$('#seg-sat-value').val()) // 4 digit representation
+        this.model.get('refEvidencePair');
+        
+        if (this.intention.getFuncSegments().length >= 2 && this.hasUD == false) {
+            if (this.intention.get('evolvingFunction').get('type') !== 'SD' && this.intention.get('evolvingFunction').get('type') !== 'DS') {
+             this.intention.getFuncSegments()[1].set('refEvidencePair', this.$('#seg-sat-value').val());
```
Comment (YesugenB) [inline leaf-ui/rappid-extensions/ElementInspector.js]: Indent

File: leaf-ui/rappid-extensions/ElementInspector.js [hunk 31]
```diff
@@ -991,6 +613,473 @@ var ElementInspector = Backbone.View.extend({
...
+        if (functionType == 'R') {
+            this.$('#seg-sat-value').html(this.satValueOptionsNoRandom());
+            this.$("#seg-sat-value").html('<option value="(no value)"> (no value) </option>');
+            this.$("#seg-sat-value").val("(no value)");
+            this.model.get('refEvidencePair');
+        } else if (this.intention.get('evolvingFunction').get('type') !== 'MP' && this.intention.get('evolvingFunction').get('type') !== 'MN' && this.intention.get('evolvingFunction').get('type') !== 'SD' && this.intention.get('evolvingFunction').get('type') !== 'DS' && functionType == 'C') {
+            this.$("#seg-sat-value").val(this.initSatValue);
+            this.model.set('refEvidencePair', this.initSatValue);
+        }
+        else if (functionType == 'I') {
```
Comment (YesugenB) [inline leaf-ui/rappid-extensions/ElementInspector.js]: Pull in the `else if` so it matches otherss

File: leaf-ui/rappid-extensions/ElementInspector.js [hunk 32]
```diff
@@ -991,6 +600,470 @@ var ElementInspector = Backbone.View.extend({
...
+                this.$("#seg-sat-value").val("0011"); 
+            } else {
+                this.$("#seg-sat-value").html(this.satValueOptionsPositiveOrNegative(prevVal, false));
+                this.$("#seg-sat-value").val("1100");
+            }
+        } else if (func == 'R') {
+            this.$("#seg-sat-value").html(this.satValueOptionsAll());
+            this.$("#seg-sat-value").val("(no value)");
+            this.$("#seg-sat-value").prop('disabled', true);
+
```
Comment (YesugenB) [inline leaf-ui/rappid-extensions/ElementInspector.js]: extra line

File: leaf-ui/rappid-extensions/ElementInspector.js [hunk 33]
```diff
@@ -991,6 +600,470 @@ var ElementInspector = Backbone.View.extend({
...
+     * 
+     * @param {String} currentValue 
+     * @param {Boolean} positive If true - increasing, if false, decreasing
+     * @returns HTML string of options with values
+     */
+    satValueOptionsPositiveOrNegative: function (currentVal, postive) {
+        var satVals = ["0011", "0010", "0000", "0100", "1100"];
+        var result = '';
+
+        if (postive) {
```
Comment (YesugenB) [inline leaf-ui/rappid-extensions/ElementInspector.js]: postive????

File: leaf-ui/rappid-extensions/ElementInspector.js [hunk 34]
```diff
@@ -991,6 +600,470 @@ var ElementInspector = Backbone.View.extend({
...
+     * This function takes in an initial value
+     * And returns an html string options with values 
+     * That are either larger or smaller than the initial value
+     * Depending on the positive boolean parameter
+     * 
+     * @param {String} currentValue 
+     * @param {Boolean} positive If true - increasing, if false, decreasing
+     * @returns HTML string of options with values
+     */
+    satValueOptionsPositiveOrNegative: function (currentVal, postive) {
```
Comment (YesugenB) [inline leaf-ui/rappid-extensions/ElementInspector.js]: postive?

File: leaf-ui/rappid-extensions/ElementInspector.js (hunk 28 above)
Comment (i-m-foster) [reply leaf-ui/rappid-extensions/ElementInspector.js]: srry i cant spell :(

**Review by amgrubb — APPROVED** [review]
Spectacular! Look through your changes and add comment to all functions where appropriate (especially in `ElementInspector.js`). When complete `Squash and Merge`.

**Comment by amgrubb** [comment]
Please don't add 14 more commits to make it to 100 again. :wink:

**Review by YesugenB — APPROVED** [review]

**Review by Domily11 — APPROVED** [review]

**Review by angelac2023 — APPROVED** [review]

**Pull request merged**